### PR TITLE
Emit close event once we reach the end of the stream

### PR DESCRIPTION
**Pull request by WyriHaximus**

### CODE REVIEW

File: src/Stream/ReadableStreamTrait.php [hunk 1]
```diff
@@ -102,6 +102,9 @@ protected function performRead($chunkSize)
                 $this->emit('end', [
                     $this,
                 ]);
+                $this->emit('close', [
+                    $this,
+                ]);
```
Comment (clue) [inline src/Stream/ReadableStreamTrait.php]: Makes sense. Maybe call `close()` instead here? Could also use a test or two :+1:

Comment (WyriHaximus) [reply src/Stream/ReadableStreamTrait.php]: Replacing both lines with close and calling close instead. Will also add tests to be sure everything works as intended :+1:

Comment (WyriHaximus) [reply src/Stream/ReadableStreamTrait.php]: Solved the issue, working on tests now. (It required some other related code change as well.)

Comment (WyriHaximus) [reply src/Stream/ReadableStreamTrait.php]: PR updated

File: src/Stream/ReadableStreamTrait.php [hunk 2]
```diff
@@ -99,9 +99,7 @@ protected function performRead($chunkSize)
             if ($this->readCursor < $this->size) {
                 $this->readChunk();
             } else {
-                $this->emit('end', [
-                    $this,
-                ]);
+                $this->close();
```
Comment (clue) [inline src/Stream/ReadableStreamTrait.php]: I'm not sure I follow exactly what' supposed to happen here, but it's my understanding that this should first emit an `end` event immediately followed by a `close`? (https://github.com/reactphp/stream#end-event)

Do we have a test for this?

Comment (WyriHaximus) [reply src/Stream/ReadableStreamTrait.php]: Good point, end shouldn't be emitted from the close call. But from when we reach EOF. Adjusted code and tests to ensure the behavior is as described on https://github.com/reactphp/stream#end-event

Comment (clue) [reply src/Stream/ReadableStreamTrait.php]: Thank you for looking into this, changes LGTM! :+1: 

For the reference: The additional `$this` arguments have been removed via https://github.com/reactphp/stream/pull/69 a while back. This component still targets legacy stream v0.4, so I guess we may want to address this eventually.

**Comment by WyriHaximus** [comment]
Cool, I'll squash and merge soon :+1:

**Pull request merged**